rust/rbac-registration/src/cardano/cip509/types/role_data_record.rs

```diff
@@ -0,0 +1,129 @@
+//! Role data record where each role data fields are stored.
+
+use std::collections::HashMap;
+
+use c509_certificate::c509::C509;
+use ed25519_dalek::VerifyingKey;
+use pallas::ledger::addresses::ShelleyAddress;
+use x509_cert::Certificate as X509;
+
+use super::{PointData, PointTxnIdx};
+use crate::cardano::cip509::extract_key;
+
+/// Actual data of key local ref. Containing X509, C509 and public key.
+#[derive(Debug, Clone)]
+pub enum CertOrPk {
+    /// X509 certificate, None if deleted.
+    X509(Option<Box<X509>>),
+    /// C509 certificate, None if deleted.
+    C509(Option<Box<C509>>),
+    /// Public key, None if deleted.
+    PublicKey(Option<VerifyingKey>),
+}
+
+impl CertOrPk {
+    /// Extract public key from the given certificate or public key.
+    pub(crate) fn extract_pk(&self) -> Option<VerifyingKey> {
+        match self {
+            CertOrPk::X509(Some(x509)) => extract_key::x509_key(x509).ok(),
+            CertOrPk::C509(Some(c509)) => extract_key::c509_key(c509).ok(),
+            CertOrPk::PublicKey(pk) => *pk,
+            _ => None,
+        }
+    }
+}
+
+/// Role data record where each field has it own record of its value and its associated
+/// point and transaction index. If a field has key rotation, then the vector index is
+/// used. Eg. Accessing key rotation 0 can be done by `signing_keys.first()`
+#[derive(Debug, Clone)]
+pub struct RoleDataRecord {
```

Comment:
Better to move this structure into its own `rs` rile

Reply:
I assume you mean the `CertOrPk`?